### PR TITLE
[core] don’t use stale collision tile in feature queries

### LINKED ISSUE #8289: Crash with FeatureIndex::addFeature while Map::queryPointAnnotations
When looking to reproduce #8202 / #8275 by constantly calling Map::queryPointAnnotation(), I was able to produce the following symbolicated stacktrace:

```cpp
********** Crash dump: **********
Build fingerprint: 'google/angler/angler:7.1.2/NPG05F/3729850:user/release-keys'
pid: 25377, tid: 25377, name: pboxsdk.testapp  >>> com.mapbox.mapboxsdk.testapp <<<
signal 7 (SIGBUS), code 1 (BUS_ADRALN), fault addr 0x1000d
Stack frame #00 pc 0015d0bc  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >::__get_long_size() const at /Users/Nurbot/mapbox-gl-native/mason_packages/osx-x86_64/android-ndk/arm-v7-9-r13b/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/string:1797
Stack frame #01 pc 0015d015  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine AnnotationTileFeature at /Users/Nurbot/mapbox-gl-native/build/android-arm-v7-9/Debug/../../../src/mbgl/annotation/annotation_tile.hpp:27
Stack frame #02 pc 0016a461  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::FeatureIndex::addFeature(std::__ndk1::unordered_map<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::vector<mapbox::geometry::feature<double>, std::__ndk1::allocator<mapbox::geometry::feature<double> > >, std::__ndk1::hash<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > >, std::__ndk1::equal_to<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > >, std::__ndk1::allocator<std::__ndk1::pair<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const, std::__ndk1::vector<mapbox::geometry::feature<double>, std::__ndk1::allocator<mapbox::geometry::feature<double> > > > > >&, mbgl::IndexedSubfeature const&, mbgl::GeometryCoordinates const&, std::experimental::fundamentals_v1::optional<std::__ndk1::vector<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::allocator<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > > > > const&, mbgl::GeometryTileData const&, mbgl::CanonicalTileID const&, mbgl::style::Style const&, float, float) const at /Users/Nurbot/mapbox-gl-native/src/mbgl/geometry/feature_index.cpp:113 (discriminator 1)
Stack frame #03 pc 0016a363  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::FeatureIndex::query(std::__ndk1::unordered_map<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::vector<mapbox::geometry::feature<double>, std::__ndk1::allocator<mapbox::geometry::feature<double> > >, std::__ndk1::hash<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > >, std::__ndk1::equal_to<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > >, std::__ndk1::allocator<std::__ndk1::pair<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const, std::__ndk1::vector<mapbox::geometry::feature<double>, std::__ndk1::allocator<mapbox::geometry::feature<double> > > > > >&, mbgl::GeometryCoordinates const&, float, double, double, std::experimental::fundamentals_v1::optional<std::__ndk1::vector<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::allocator<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > > > > const&, mbgl::GeometryTileData const&, mbgl::CanonicalTileID const&, mbgl::style::Style const&, mbgl::CollisionTile const*) const at /Users/Nurbot/mapbox-gl-native/src/mbgl/geometry/feature_index.cpp:90
Stack frame #04 pc 001d2075  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::GeometryTile::queryRenderedFeatures(std::__ndk1::unordered_map<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::vector<mapbox::geometry::feature<double>, std::__ndk1::allocator<mapbox::geometry::feature<double> > >, std::__ndk1::hash<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > >, std::__ndk1::equal_to<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > >, std::__ndk1::allocator<std::__ndk1::pair<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const, std::__ndk1::vector<mapbox::geometry::feature<double>, std::__ndk1::allocator<mapbox::geometry::feature<double> > > > > >&, mbgl::GeometryCoordinates const&, mbgl::TransformState const&, std::experimental::fundamentals_v1::optional<std::__ndk1::vector<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::allocator<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > > > > const&) at /Users/Nurbot/mapbox-gl-native/src/mbgl/tile/geometry_tile.cpp:148
Stack frame #05 pc 000b8e73  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::style::Source::Impl::queryRenderedFeatures(mbgl::style::QueryParameters const&) const at /Users/Nurbot/mapbox-gl-native/src/mbgl/style/source_impl.cpp:246
Stack frame #06 pc 000bd071  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::style::Style::queryRenderedFeatures(mbgl::style::QueryParameters const&) const at /Users/Nurbot/mapbox-gl-native/src/mbgl/style/style.cpp:523
Stack frame #07 pc 0009d5a9  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::Map::queryRenderedFeatures(mapbox::geometry::box<double> const&, std::experimental::fundamentals_v1::optional<std::__ndk1::vector<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::allocator<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > > > > const&) at /Users/Nurbot/mapbox-gl-native/src/mbgl/map/map.cpp:826 (discriminator 1)
Stack frame #08 pc 0009d6d3  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::Map::queryPointAnnotations(mapbox::geometry::box<double> const&) at /Users/Nurbot/mapbox-gl-native/src/mbgl/map/map.cpp:840 (discriminator 3)
Stack frame #09 pc 00073283  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine mbgl::android::NativeMapView::queryPointAnnotations(_JNIEnv&, jni::Object<mbgl::android::RectF>) at /Users/Nurbot/mapbox-gl-native/platform/android/src/native_map_view.cpp:704 (discriminator 2)
Stack frame #10 pc 00088193  /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm/libmapbox-gl.so: Routine operator()<jni::jobject *, jni::jobject *> at /Users/Nurbot/mapbox-gl-native/build/android-arm-v7-9/Debug/../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:59
Stack frame #11 pc 00525705  /data/app/com.mapbox.mapboxsdk.testapp-2/oat/arm/base.odex (offset 0x4fc000)
```

Been trying to produce this with a core gl test, but so far no luck.

### PR DESCRIPTION
**Pull request by ivovandongen**
fixes #8289

We could make this check more granular since the `FeatureIndex` is still usable in this case, but since it's an edge case, I don't know if it's worth it.

### CODE REVIEW

**Comment by ivovandongen** [comment]
To be able to stub out the `FeatureIndex` for testing I had to make [FeatureIndex#query](https://github.com/mapbox/mapbox-gl-native/blob/cdd00aa27e2edc1767dc9fa77ce3ebe93b7f8661/src/mbgl/geometry/feature_index.hpp#L38) virtual

File: src/mbgl/tile/geometry_tile.cpp [hunk 1]
```diff
@@ -147,7 +147,7 @@ void GeometryTile::queryRenderedFeatures(
     const TransformState& transformState,
     const RenderedQueryOptions& options) {
 
-    if (!featureIndex || !data) return;
+    if (!featureIndex || !data || availableData != DataAvailability::All) return;
```
Comment (jfirebaugh) [inline src/mbgl/tile/geometry_tile.cpp]: Instead of returning an empty result set for queries made between layout and placement, I think we could effectively exclude symbol layers from the query by resetting `collisionTile` in `GeometryTile::onLayout`. Then checking the collision tile [will be skipped](https://github.com/mapbox/mapbox-gl-native/blob/bce352bd86f7faf98b3ba5bbb31d3a7860533199/src/mbgl/geometry/feature_index.cpp#L85-L88), but querying everything else is still possible.

File: src/mbgl/annotation/annotation_tile.hpp [hunk 2]
```diff
@@ -45,7 +45,10 @@ class AnnotationTileLayer : public GeometryTileLayer {
     AnnotationTileLayer(std::string);
 
     std::size_t featureCount() const override { return features.size(); }
-    std::unique_ptr<GeometryTileFeature> getFeature(std::size_t i) const override { return std::make_unique<AnnotationTileFeature>(features[i]); }
+    std::unique_ptr<GeometryTileFeature> getFeature(std::size_t i) const override {
+        assert(i < features.size());
+        return std::make_unique<AnnotationTileFeature>(features[i]);
```
Comment (jfirebaugh) [inline src/mbgl/annotation/annotation_tile.hpp]: Instead of adding an assertion, let's change `features[i]` to `features.at(i)`. The latter has bounds-checking built in, and is what `VectorTileLayer::getFeature` [uses](https://github.com/mapbox/mapbox-gl-native/blob/bce352bd86f7faf98b3ba5bbb31d3a7860533199/src/mbgl/tile/vector_tile.cpp#L313).

File: src/mbgl/geometry/feature_index.hpp [hunk 3]
```diff
@@ -35,7 +35,7 @@ class FeatureIndex {
 
     void insert(const GeometryCollection&, std::size_t index, const std::string& sourceLayerName, const std::string& bucketName);
 
-    void query(
+    virtual void query(
```
Comment (jfirebaugh) [inline src/mbgl/geometry/feature_index.hpp]: Here's an alternative way to write the test that uses a real `FeatureIndex` object. Still requires a lot of setup, but I think using real objects as much as possible is good.

```C++
TEST(VectorTile, Issue8289) {
    VectorTileTest test;
    VectorTile tile(OverscaledTileID(0, 0, 0), "source", test.updateParameters, test.tileset);

    auto data = std::make_unique<AnnotationTileData>();
    data->layers.emplace("test", AnnotationTileLayer("test"));
    data->layers.at("test").features.push_back(AnnotationTileFeature(0, FeatureType::Point, GeometryCollection()));

    // Simulate layout and placement of a symbol layer.
    tile.onLayout(GeometryTile::LayoutResult {
        {},
        std::make_unique<FeatureIndex>(),
        std::move(data),
        0
    });

    auto collisionTile = std::make_unique<CollisionTile>(PlacementConfig());

    IndexedSubfeature subfeature { 0, "", "", 0 };
    CollisionFeature feature(GeometryCoordinates(), Anchor(0, 0, 0, 0), -5, 5, -5, 5, 1, 0, style::SymbolPlacementType::Point, subfeature, false);
    collisionTile->insertFeature(feature, 0, true);
    collisionTile->placeFeature(feature, false, false);

    tile.onPlacement(GeometryTile::PlacementResult {
        {},
        std::move(collisionTile),
        0
    });

    // Simulate a second layout with empty data.
    tile.onLayout(GeometryTile::LayoutResult {
        {},
        std::make_unique<FeatureIndex>(),
        std::make_unique<AnnotationTileData>(),
        0
    });

    std::unordered_map<std::string, std::vector<Feature>> result;
    GeometryCoordinates queryGeometry {{ Point<int16_t>(0, 0) }};
    TransformState transformState;
    RenderedQueryOptions options;

    tile.queryRenderedFeatures(result, queryGeometry, transformState, options);

    EXPECT_TRUE(result.empty());
}
```

I verified that in current `master`, this fails -- not at the same point as in #8289, but enough to show that a data synchronization issue is present, and either of the proposed fixes will solve it.

File: src/mbgl/annotation/annotation_tile.hpp (hunk 2 above)
Comment (ivovandongen) [inline src/mbgl/annotation/annotation_tile.hpp]: Check. I've changed this, but since `vector::at` throws an exception, which will still crash the app, I'm wondering if we shouldn't catch, log an error and continue in these cases. At least in release builds.

**Comment by ivovandongen** [comment]
Thanks @jfirebaugh. I've made the requested changes, adding the test in a separate file for clarity.

File: src/mbgl/annotation/annotation_tile.hpp (hunk 2 above)
Comment (jfirebaugh) [inline src/mbgl/annotation/annotation_tile.hpp]: For logic errors, it's better to crash -- crashes are picked up by reporting tools, we get a bug report, and can fix the issue. In comparison, logs are mostly ignored.
